Add Monix interop layer

mijicd opened the pull request:
This PR resolves #395.

## Contents

- added conversions to and from `monix.eval.Task`
- added conversions to and from `monix.eval.Coeval`
- bumped monix to `3.0.0-RC2`
- updated tut documentation

## Remarks

I am uncertain whether or not to provide conversions for `monix.eval.Fiber`. While it is specified in ticket description, there are a couple of confusing points:

- it is not a "top-level" primitive Monix exposes
- there's nothing similar to ZIO fiber's `poll` (return immediately if not completed)

mijicd commented:
@jdegoes test suite for 2.12 fails with:

```
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received
```

The complete log can be found [here](https://travis-ci.org/scalaz/scalaz-zio/jobs/472373206). 

Since I don't have write permissions, could you take a look at this and advise?

jdegoes commented:
@mijicd This indicates a hanging test. Most of the tests in `core` are protected against this because they extend `AbstractRTSSpec`, which times out all runaway tests. However, the tests in the interop packages are not so protected (and there are a few ScalaCheck / ScalaTest checks inside `core` that do not have such protections).

The main thing to look at is to figure out which test suite is running when it hangs. You can also turn on incremental output in SBT so you can see the last line of text printed before the hanging test.

Finally, occasionally a test seems to hang for unrelated reasons. I'll restart the build in case that's what happened.

mijicd commented:
@jdegoes I’ll take a look. It is still a “smell” since the tests I introduced inherit AbstractRTSSpec, so they shouldn’t hang.

> On Dec 26, 2018, at 5:31 PM, John A. De Goes <notifications@github.com> wrote:
> 
> @mijicd This indicates a hanging test. Most of the tests in core are protected against this because they extend AbstractRTSSpec, which times out all runaway tests. However, the tests in the interop packages are not so protected (and there are a few ScalaCheck / ScalaTest checks inside core that do not have such protections).
> 
> The main thing to look at is to figure out which test suite is running when it hangs. You can also turn on incremental output in SBT so you can see the last line of text printed before the hanging test.
> 
> Finally, occasionally a test seems to hang for unrelated reasons. I'll restart the build in case that's what happened.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

mijicd commented:
@jdegoes it seems that `catzSpec` is the one causing hanging. Further inspection made me realize that this is one of the tests not using `AbstractRTSSpec` and `specs2` at all. I'll play with it and try to fix the hanging.

mijicd commented:
A couple updates:
- replaced `unsafeRunSync` with `unsafeRunTimed` using timeout od 10 seconds
- updated `cats-effect` and `cats-effect-laws` to `1.1.0`
- tests passed once `ConcurrentTests` were disabled
- failures can't be replicated locally

Finally, the last printed line before hanging for 10 mins and eventually failing tests was:

```
[info] Fast optimizing /home/travis/build/scalaz/scalaz-zio/interop-cats/js/target/scala-2.12/scalaz-zio-interop-cats-test-fastopt.js
```

jdegoes commented:
@mijicd This is clearly not related to the Monix interop or tests you added. Must be an existing problem. So let's go ahead and get this merged when you're ready!

mijicd commented:
@jdegoes Thanks! I'm starting with ZIO instance(s) and then when I'm done we can proceed with merge.

mijicd commented:
@jdegoes the problem I see with providing `Task`/`Coeval` instances is that neither of them are bifunctors, which is implied by `liftZIO` signature:

```scala
def liftZIO[E, A](io: IO[E, A]): F[E, A]
```

If there's a workaround this I'd be happy to implement it. Otherwise, let's wait for Travis (seems to be pending for awhile now), and merge this.

jdegoes commented:
Let's just merge this now (I can create tickets for the other). In general, the solution is to embed the `E` into a throwable, or to define the instance for `EitherT` (which suggests it should be done in the cats project, not here).

mijicd commented:
Agreed. As soon as the build is started and green it can be merged.
